mitiq/parameter.py

```diff
@@ -0,0 +1,257 @@
+            CZPowGate]
+
+@converter
+def scale_parameters(
+    circ: Circuit,
+    scale_factor: float,
+    sigma: float,
+    seed: int = None
+) -> Circuit:
+    """Adds parameter noise to a circuit with level noise.
+    This adds noise to the actual parameter instead of
+    adding an parameter channel.
+
+    Args:
+        circ: The quantum program as a Cirq circuit object. All measurements
+            should be in the last moment of the circuit.
+        scale_factor: Amount to scale the base noise level of parameter rotations by.
+        sigma: Base noise level (variance) in parameter rotations
+        seed: 
+
+    Returns:
+        The input circuit with scaled rotation angles
+
+    """
+    final_moments = []
+    noise = (scale_factor - 1) * sigma
+    for moment in circ:
+        curr_moment = []
+        for op in moment.operations:
+            gate = copy.deepcopy(op.gate)
+            qubits = op.qubits
+            if isinstance(gate, MeasurementGate):
+                curr_moment.append(gate(*qubits))
+            else:
+                base_gate = _get_base_gate(gate)
+                param = gate.exponent * np.pi
+                rng = np.random.RandomState(seed)
+                error = rng.normal(loc=0.0, scale=np.sqrt(noise))
+                new_param = (param + error)
+                curr_moment.append(
+                    base_gate(exponent=new_param/np.pi)(*qubits))
+        final_moments.append(Moment(curr_moment))
+    return Circuit(final_moments)
+
+
+def _get_base_gate(gate):
+    for base_gate in BASE_GATES:
+        if isinstance(gate, base_gate):
+            return base_gate
+    raise Exception(
+        "Must have circuit be made of rotation gates. \
+        Your gate {} may not be supported".format(gate))
+
+
+def scale_parameters_channel(circ: Circuit, noise=None) -> Circuit:
+    """Adds parameter noise to a circuit with level noise via noise channel.
+
+    Args:
+        circ: The quantum program as a cirq object.
+        noise: the base noise level to put into circuit (variance of rotations)
+
+    Returns:
+        Noisy circuit with noise gates after each appropriate gate.
+
+    """
+    final_moments = []
+    for moment in circ:
+        final_moments.append(moment)
+        curr_moment = []
+        for op in moment.operations:
+            gate = op.gate
+            qubits = op.qubits
+            if len(qubits) == 1:
+                qubits = qubits[0]
+                noisy_op = parameter_noise_1q(noise, gate)(qubits)
+            elif len(qubits) == 2:
+                noisy_op = parameter_noise_2q(noise, gate)(qubits[0], qubits[1])
+            else:
+                raise Exception(
+                    "Gates for more than two qubits are not \
+                    supported with parameter noise scaling.")
+            curr_moment.append(noisy_op)
+        final_moments.append(Moment(curr_moment))
+    return Circuit(final_moments)
```

Comment:
Do I understand right that `scale_parameters` is the scaling method that users are going to use, while `scale_parameters_channel` is used to model the noise of a parameter noise backend for testing?

`scale_parameters` is multi-backend compatible and so seems better than the `channel` method.

@yhindy  can you say a bit more about the use case for `scale_parameters_channel`?

Reply:
yeah scale parameters channel is for simulation on density matrix solvers. the `scale_parameters` is multi-backend compatible. i think we discussed this and the `channel` one should be removed from `mitiq` proper.